packages/core/src/mcp/oauth-provider.ts

```diff
@@ -0,0 +1,789 @@
+      const authServerMetadataUrl = new URL(
+        '/.well-known/oauth-authorization-server',
+        serverUrl,
+      ).toString();
+
+      const authServerResponse = await fetch(authServerMetadataUrl);
+      if (!authServerResponse.ok) {
+        throw new Error(
+          'Failed to fetch authorization server metadata for client registration',
+        );
+      }
+
+      const authServerMetadata = await authServerResponse.json();
+
+      // Register client if registration endpoint is available
+      if (authServerMetadata.registration_endpoint) {
+        const clientRegistration = await this.registerClient(
+          authServerMetadata.registration_endpoint,
+          config,
+        );
+
+        config.clientId = clientRegistration.client_id;
+        if (clientRegistration.client_secret) {
+          config.clientSecret = clientRegistration.client_secret;
+        }
+
+        console.log('Dynamic client registration successful');
+      } else {
+        throw new Error(
+          'No client ID provided and dynamic registration not supported',
+        );
+      }
+    }
+
+    // Validate configuration
+    if (!config.clientId || !config.authorizationUrl || !config.tokenUrl) {
+      throw new Error(
+        'Missing required OAuth configuration after discovery and registration',
+      );
+    }
+
+    // Generate PKCE parameters
+    const pkceParams = this.generatePKCEParams();
+
+    // Build authorization URL
+    const authUrl = this.buildAuthorizationUrl(config, pkceParams);
+
+    console.log('\nOpening browser for OAuth authentication...');
+    console.log('If the browser does not open, please visit:');
+    console.log(authUrl);
+
+    // Start callback server
+    const callbackPromise = this.startCallbackServer(pkceParams.state);
+
+    // Open browser
+    try {
+      await open(authUrl);
+    } catch (error) {
+      console.warn(
+        'Failed to open browser automatically:',
+        getErrorMessage(error),
+      );
+    }
+
+    // Wait for callback
+    const { code } = await callbackPromise;
+
+    console.log('\nAuthorization code received, exchanging for tokens...');
+
+    // Exchange code for tokens
+    const tokenResponse = await this.exchangeCodeForToken(
+      config,
+      code,
+      pkceParams.codeVerifier,
+    );
+
+    // Convert to our token format
+    const token: MCPOAuthToken = {
+      accessToken: tokenResponse.access_token,
+      tokenType: tokenResponse.token_type,
+      refreshToken: tokenResponse.refresh_token,
+      scope: tokenResponse.scope,
+    };
+
+    if (tokenResponse.expires_in) {
+      token.expiresAt = Date.now() + tokenResponse.expires_in * 1000;
+    }
+
+    // Save token
+    try {
+      await MCPOAuthTokenStorage.saveToken(
+        serverName,
+        token,
+        config.clientId,
+        config.tokenUrl,
+      );
+      console.log('Authentication successful! Token saved.');
```

Comment:
May want to consider debug calls for a lot of these. I'll stop leaving the comment of log -> debug on the PR though